File: 67-add-binary/solution.cpp
#include <string>
#include <deque>

class Solution
{
public:
  std::string addBinary(std::string a, std::string b)
  {
    std::deque<char> dq;
    int a_pointer = a.size() - 1, b_pointer = b.size() - 1;
    bool leftover = false;
    while (a_pointer >= 0 || b_pointer >= 0 || leftover)
    {
      const char a_bit = a_pointer < 0 ? '0' : a[a_pointer], b_bit = b_pointer < 0 ? '0' : b[b_pointer];
      --a_pointer;
      --b_pointer;

      int ones = 0;
      if (a_bit == '1')
      {
        ++ones;
      }

      if (b_bit == '1')
      {
        ++ones;
      }

      if (leftover)
      {
        ++ones;
      }

      leftover = ones >= 2;
      dq.emplace_front(ones % 2 == 0 ? '0' : '1');
    }

    std::string result(dq.begin(), dq.end());
    return result;
  }
};

File: 67-add-binary/solution.py
from collections import deque


class Solution:
  def addBinary(self, a: str, b: str) -> str:
    dq: deque[str] = deque()
    a_pointer, b_pointer, leftover = len(a) - 1, len(b) - 1, False
    while a_pointer >= 0 or b_pointer >= 0 or leftover:
      a_bit = '0' if a_pointer < 0 else a[a_pointer]
      b_bit = '0' if b_pointer < 0 else b[b_pointer]

      a_pointer -= 1
      b_pointer -= 1

      ones = 0
      for bit in [a_bit, b_bit]:
        if bit == '1':
          ones += 1

      if leftover:
        ones += 1

      leftover = ones >= 2
      dq.appendleft('0' if ones % 2 == 0 else '1')

    return ''.join(dq)
